xarray/core/indexing.py

```diff
@@ -444,25 +427,49 @@ def orthogonally_indexable(array):
     return array
 
 
-class NumpyIndexingAdapter(utils.NDArrayMixin):
-    """Wrap a NumPy array to use orthogonal indexing (array indexing
-    accesses different dimensions independently, like netCDF4-python variables)
+def _outer_to_numpy_indexer(key, shape):
+    """Convert an OuterIndexer into an indexer for NumPy.
+
+    Parameters
+    ----------
+    key : OuterIndexer
+        Outer indexing tuple to convert.
+    shape : tuple
+        Shape of the array subject to the indexing.
+
+    Returns
+    -------
+    tuple
+        Base tuple suitable for use to index a NumPy array.
     """
-    # note: this object is somewhat similar to biggus.NumpyArrayAdapter in that
-    # it implements orthogonal indexing, except it casts to a numpy array,
-    # isn't lazy and supports writing values.
-    def __init__(self, array):
-        self.array = np.asarray(array)
+    if len([k for k in key if not isinstance(k, slice)]) <= 1:
+        # If there is only one vector and all others are slice,
+        # it can be safely used in mixed basic/advanced indexing.
+        # Boolean index should already be converted to integer array.
+        return tuple(key)
 
-    def __array__(self, dtype=None):
-        return np.asarray(self.array, dtype=dtype)
+    n_dim = len([k for k in key if not isinstance(k, integer_types)])
+    i_dim = 0
+    new_key = []
+    for k, size in zip(key, shape):
+        if isinstance(k, integer_types):
+            new_key.append(k)
+        else:  # np.ndarray or slice
+            if isinstance(k, slice):
+                k = np.arange(*k.indices(size))
+            assert k.dtype.kind == 'i'
```

Comment:
@fujiisoup - not 100% sure but is it possible that this should be `assert k.dtype.kind in {'i', 'u'}`? See regression tests in #1544 .

Reply:
Thanks for finding the point.
I think it will work but I am not yet 100% sure a single change of this line totally fixes #1544. 
I will investigate more inside later today.